h2o-py/h2o/utils/shared_utils.py

```diff
@@ -567,3 +568,12 @@ def __exit__(self, exc_type, exc_value, traceback):
         if self._file_name is None:
             return
         self.write_to_file(self._file_name)
+
+
+@contextlib.contextmanager
+def as_resource(res):
+    try:
+        yield res
+    finally:
+        if hasattr(res, "close") and callable(res.close):
+            res.close()
```

Comment:
works:
```python
In [2]: val = 456
In [3]: with as_resource(val) as v:
    ...:     print(v+123)
    ...:
579
```
and
```python
In [25]: file = open("/Users/seb/Downloads/KDDCup09_appetency_infogram.csv")
In [26]: with as_resource(file) as f:
    ...:     lines = f.readlines()
    ...: print(lines[:5])
['"Features","Relevance","CMI"\n', '"Var192",1.0,1.0\n', '"Var197",0.7313172543235006,0.7761696052436258\n', '"Var126",0.3606564328906574,0.537418539624436\n', '"Var204",0.46322020333322306,0.09712901681870682\n']

In [27]: f.closed
Out[27]: True

In [28]: file.closed
Out[28]: True
```

and even better, thanks to `abc` magic
```python
In [29]: class A:
    ...:     def __enter__():
    ...:         pass
    ...:     def __exit__():
    ...:         pass
    ...:

In [30]: a = A()

In [31]: isinstance(a, contextlib.AbstractContextManager)
Out[31]: True
```
so all contextmanagers will be correctly detected by the suggestion above

Reply:
Well, I could check for `__enter__` and `__exit__`. I was too concentrated on closing the resource which doesn't really have to happen in `__exit__` that's why we have the [closing](https://docs.python.org/3/library/contextlib.html#contextlib.closing) context manager that makes sure it will `close` the thing. But for files it should work both ways.

Reply:
for Py<3.6, this addition to `compatibility` class does the trick:

```python
try:
    from contextlib import AbstractContextManager
except ImportError:
    import abc
    from future.utils import with_metaclass
    class AbstractContextManager(with_metaclass(abc.ABCMeta)):
         @classmethod
         def __subclasshook__(cls, C):
              if cls is AbstractContextManager:
                 return all(any(m in SC.__dict__ for SC in C.__mro__) for m in ("__enter__", "__exit__"))
              return NotImplemented


@contextmanager
def as_resource(o):
    if isinstance(o, AbstractContextManager): 
        with(o) as res:
            yield res
    else:
        yield o
```

Reply:
Awesome! Thank you!